text/0193-removal-of-type-unions.md

```diff
@@ -0,0 +1,575 @@
+includes two kinds of APIs:
+
+- [Generated as part of the CloudFormation Resources](#generated-code)
+- [Manually authored as part of the AWS Construct Library or AWS CDK Core framework](#hand-written-code)
+
+# Detailed Design
+
+## Generated code
+
+As previously discussed, the _CloudFormation Resources_ ubiquitously rely on _type unions_ to allow providing late-bound
+values to the various properties accepted by the resources. Late-binding values is necessary to support mutation APIs
+which provide enhanced ergonomics for certain constructs, particularly where various components can be expected to
+collaborate on the finalized configuration of a resource.
+
+### Where existing APIs already support the use-case in a safe manner
+
+All _type unions_ where late-bound values can be represented in the literal type of the value (`string`, `number` and
+`any`) can be simplified to that literal type, since the required APIs already exist and the migration path for those
+use-cases is already very well defined: using `Lazy` factory methods to manually wrap `IResolvable` instances:
+- Current instances of:
+  ```ts
+  public propertyName: string | IResolvable;
+  ```
+
+- Are changed to:
+  ```ts
+  public propertyName: string;
+  ```
+
+- And places where an `IResolvable` was passed (if any) are replaced to:
+  ```ts
+  let value: IResolvable;
+  // ...
+  { propertyName: Lazy.stringValue(value) }
+  ```
+
+### Use-cases where existing APIs are not usable from statically typed languages
+
+On the other hand, _type unions_ of `IResolvable` and complex types (including collections) cannot be resolved that
+easily. In **TypeScript** and other dynamic languages (such as **Python**), the `Lazy.asAny` method can be used to wrap
+an `IResolvable` and evade the type checker. This approach is however no workable in **C#**, **Go** and **Java** (and
+likely other statically typed languages we may wish to support in the future). For example the `s3.CfnBucket` accepts
+(and exposes) properties such as:
+```ts
+lifecycleConfiguration: CfnBucket.LifecycleConfigurationProperty | cdk.IResolvable | undefined;
+```
+
+#### Recommended Approach
+
+For those cases, one solution is to have users of statically typed languages use _[escape hatches]_ to forcefully
+override properties with the desired late-bound value, since _[escape hatches]_ offer `any`-typed accessors. This
+results in a degradation of ergonomics for statically typed languages, but not not imply a redution in available
+features offered by the AWS CDK:
+
+- Where **TypeScript** developers could use:
+  ```ts
+  let cfnBucket: CfnBucket;
+  let value: IResolvable;
+  // ...
+  bucket.lifecycleConfiguration = value as any;
+  ```
+
+- Developers in **Java** (and other statically typed languages) would instead use:
+  ```java
+  CfnBucket bucket;
+  IResolvable resolvable;
+  // ...
+  bucket.addPropertyOverride('LifecycleConfiguration', resolvable);
+  ```
```

Comment:
Given that this issue is only with L1s and not the L2s are easy to fix, how many different types (for lack of a better word) of type unions exist in our current L1s? i.e., are they all union-ed with `IResolvable` vs unions are across significantly different types (`number | string`, etc.) vs unions of scalar and vector (`string | string[]`)?

Given CF has a certain level of types inherently, there can't be **that many** variations, right?

I'm wondering if we can also change our L1 generation so that this type erasure is not lost. Taking your example as an illustration -

```ts
lifecycleConfiguration: CfnBucket.LifecycleConfigurationProperty;
lifecycleConfigurationLazy: cdk.IResolvable;
```